crates/turbo-tasks/src/vc/mod.rs

```diff
@@ -282,6 +304,13 @@ where
     pub fn cell(inner: Inner) -> Self {
         <T::CellMode as VcCellMode<T>>::cell(inner)
     }
+
+    pub fn local_cell(inner: Inner) -> Self {
+        // `T::CellMode` isn't applicable here, we always create new local cells. Local
```

Comment:
Bit of a random question but just want to challenge my thinking. We have a few ways of interacting 'up the chain' so to speak. At the top level you have the global 'arena' with any number of 'subarenas' for local tasks.  Presumably you can interleave local and global tasks at will. Do we intend on having local areanas further down the call graph be able to read / access from Vcs in their 'superarenas' and global state?

```rust
#[turbo::function]
fn a(path: Vc<Path>) -> Vc<CompiledCode> {
  let is_dir = path.is_directory().await;
  if (is_dir) {
    return b(path);
  }
  // do stuff
}

#[turbo::function(local)]
fn b(path: Vc<Path>) -> Vc<CompiledCode> {
  // use path.is_directory()
}
```

Path::is_directory is already resolved in the global cache.

Reply:
> Do we intend on having local areanas further down the call graph be able to read / access from Vcs in their 'superarenas' and global state?

This is a good question.

Global state: Yes. Local tasks will create local Vcs by default, but they will be able to both create and read resolved Vcs that use global cells. When creating global cells, those cells will be assigned to nearest non-local parent task.

In your example, `Path::is_directory` would not need to be re-executed because it's in the global cache.

"Superarenas" of calling tasks: No, at least for right now. Local Vcs must be resolved before being passed as an argument into another task.

One of the unintuitive challenges with reading from local "superarenas" is that (currently) every task is `tokio::spawn`ed and never canceled/aborted. This means that a child called task can outlive it's parent caller. This isn't common as most calling tasks await all their children, but can happen in the case of panics in the parent/calling task or with use of APIs like `tokio::select!`.

For memory safety we either need a mechanism for canceling tasks (triggering a drop at the await point) if they try to read from an expired caller's "superarena", or (more likely) we need to defer the cleanup of local arenas until all of their child spawned tasks exit.